common/rpc/encryption/localStoreCertProvider.go

```diff
@@ -246,62 +264,188 @@ func (s *localStoreCertProvider) getClientTLSSettings(isWorker bool) *config.Cli
 	} else if isWorker {
 		return s.legacyWorkerSettings // legacy config case when we use Frontend.Client settings
 	} else {
+		if s.tlsSettings == nil {
+			return nil
+		}
 		return &s.tlsSettings.Client // internode client case
 	}
 }
 
-func buildCAPoolFromData(caData []string) (*x509.CertPool, error) {
-	atLeastOneCert := false
-	caPool := x509.NewCertPool()
+func (s *localStoreCertProvider) Expiring(fromNow time.Duration,
+) (expiring CertExpirationMap, expired CertExpirationMap, errs []error) {
+
+	expiring = make(CertExpirationMap)
+	expired = make(CertExpirationMap)
+	errs = make([]error, 0)
+	when := time.Now().UTC().Add(fromNow)
+
+	errs = checkTLSCertForExpiration(s.FetchServerCertificate, when, expiring, expired, errs)
+	errs = checkTLSCertForExpiration(
+		func() (*tls.Certificate, error) { return s.FetchClientCertificate(false) },
+		when, expiring, expired, errs)
+	errs = checkTLSCertForExpiration(
+		func() (*tls.Certificate, error) { return s.FetchClientCertificate(true) },
+		when, expiring, expired, errs)
+
+	// load CA certs, so that they are cached in memory
+	errs = loadCAsAndCaptureErrors(s.FetchClientCAs, errs)
+	checkCertsForExpiration(s.clientCACerts, when, expiring, expired)
+
+	errs = loadCAsAndCaptureErrors(func() (*x509.CertPool, error) {
+		return s.FetchServerRootCAsForClient(false)
+	}, errs)
+	checkCertsForExpiration(s.serverCACerts, when, expiring, expired)
+
+	errs = loadCAsAndCaptureErrors(func() (*x509.CertPool, error) {
+		return s.FetchServerRootCAsForClient(true)
+	}, errs)
+	checkCertsForExpiration(s.serverCACertsWorker, when, expiring, expired)
+
+	return expiring, expired, errs
+}
 
-	for _, ca := range caData {
-		if ca == "" {
-			continue
-		}
+func checkTLSCertForExpiration(
+	fetchCert func() (*tls.Certificate, error),
+	time time.Time,
+	expiring CertExpirationMap,
+	expired CertExpirationMap,
+	errors []error,
+) []error {
+
+	return fetchAndCheckCertForExpiration(
+		func() (*x509.Certificate, error) {
```

Comment:
[Nit] there is a LOT of nesting of anonymously defined functions, which is making this harder to read. can we extract these into named functions just to make this easier to read?